File: php/test/OneTimePasswordEntityTest.php
<?php
declare(strict_types=1);

// OneTimePassword entity test

require_once __DIR__ . '/../thesmsworks_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class OneTimePasswordEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = ThesmsworksSDK::test(null, null);
        $ent = $testsdk->OneTimePassword(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = one_time_password_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "one_time_password." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $one_time_password_ref01_ent = $client->OneTimePassword(null);
        $one_time_password_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.one_time_password"), "one_time_password_ref01"));

        $one_time_password_ref01_data_result = $one_time_password_ref01_ent->create($one_time_password_ref01_data, null);
        $one_time_password_ref01_data = Helpers::to_map($one_time_password_ref01_data_result);
        $this->assertNotNull($one_time_password_ref01_data);

        // LOAD
        $one_time_password_ref01_match_dt0 = [];
        $one_time_password_ref01_data_dt0_loaded = $one_time_password_ref01_ent->load($one_time_password_ref01_match_dt0, null);
        $this->assertNotNull($one_time_password_ref01_data_dt0_loaded);

    }
}

function one_time_password_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/one_time_password/OneTimePasswordTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = ThesmsworksSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["one_time_password01", "one_time_password02", "one_time_password03", "otp01", "otp02", "otp03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID" => $idmap,
        "THESMSWORKS_TEST_LIVE" => "FALSE",
        "THESMSWORKS_TEST_EXPLAIN" => "FALSE",
        "THESMSWORKS_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["THESMSWORKS_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["THESMSWORKS_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new ThesmsworksSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["THESMSWORKS_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["THESMSWORKS_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: cpp/test/one_time_password_entity_test.cpp
// Generated basic-flow test for the one_time_password entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct OneTimePasswordSetup {
  std::shared_ptr<ThesmsworksSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static OneTimePasswordSetup one_time_password_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/one_time_password/OneTimePasswordTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = ThesmsworksSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("one_time_password01"), Value("one_time_password02"), Value("one_time_password03"), Value("otp01"), Value("otp02"), Value("otp03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID", idmap},
    {"THESMSWORKS_TEST_LIVE", Value("FALSE")},
    {"THESMSWORKS_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "THESMSWORKS_TEST_ONE_TIME_PASSWORD_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "THESMSWORKS_TEST_LIVE") == Value("TRUE");

  OneTimePasswordSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void one_time_password_entity_instance() {
  auto testsdk = ThesmsworksSDK::testSDK();
  auto ent = testsdk->one_time_password();
  ASSERT_EQ(ent->getName(), std::string("one_time_password"), "entity name");
}

static void one_time_password_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"one_time_password", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = ThesmsworksSDK::testSDK(seed, sdkopts);
  auto se = strsdk->one_time_password();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = ThesmsworksSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->one_time_password();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void one_time_password_entity_basic() {
  auto setup = one_time_password_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : {"create", "load"}) {
    auto sk = is_control_skipped("entityOp", std::string("one_time_password.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto one_time_password_ref01_ent = client->one_time_password();
  Value one_time_password_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "one_time_password"}), "one_time_password_ref01"));
  if (!one_time_password_ref01_data.is_map()) one_time_password_ref01_data = vmap();
  {
    Value one_time_password_ref01_data_result = one_time_password_ref01_ent->create(Struct::clone(one_time_password_ref01_data), Value::undef());
    one_time_password_ref01_data = Helpers::toMapAny(one_time_password_ref01_data_result);
    if (!one_time_password_ref01_data.is_map()) one_time_password_ref01_data = vmap();
    ASSERT_TRUE(one_time_password_ref01_data.is_map(), "expected create result to be a map");
  }

  // LOAD
  Value one_time_password_ref01_match_dt0 = vmap();
  Value one_time_password_ref01_data_dt0_loaded = one_time_password_ref01_ent->load(one_time_password_ref01_match_dt0, Value::undef());
  ASSERT_TRUE(!one_time_password_ref01_data_dt0_loaded.is_undef(), "expected load result to be non-nil");

}

int main() {
  T_RUN(one_time_password_entity_instance);
  T_RUN(one_time_password_entity_stream);
  T_RUN(one_time_password_entity_basic);
  return sdktest::summary("one_time_password_entity_test");
}
